### DISCUSSION TITLE
[keyball44]REMAPは正常に動作するがQMK Firmwareでフラッシュすると左右反転する

### DISCUSSION THREAD

**Discussion by monokaai** [Yowkees/keyball#546] [Q&A] [answered]
お世話になっております。いつも便利に利用させて頂いております。
これまでREMAPを利用しており、ファームウェアは[ご配布のKeyball44_via_v1.3.2](https://remap-keys.app/catalog/tAJ9Htme4oNabUkx4832/firmware)を使っていました。

詳細な設定を試したく[リポジトリから](https://github.com/Yowkees/keyball/tree/main/qmk_firmware/keyboards/keyball/keyball44)クローン・ビルドした.hexファイルをQMK Toolboxで書き込んでみたのですが、何度やっても左右反転してしまいます（QWERTYでいうとG↔︎H、F↔︎Jのようにユニット内で線対称かつユニット自体も線対称）。

半田付けの状態は以下の写真の通りです（トラックボールは右側配置）。

![IMG_1442](https://github.com/Yowkees/keyball/assets/93644040/dab1a97a-5e62-4498-aae9-eebc419371d2)
![IMG_1443](https://github.com/Yowkees/keyball/assets/93644040/607c7d22-9124-4339-9263-db650ddc4909)
この原因と解決法についてお知恵をお借りしたいです。
初心者のため、情報に不足があればお手数ですが合わせてご指摘頂ければ幸いです。
よろしくお願いいたします。

**Comment by Yowkees** [comment]
この度はKeyball44をお選びいただき、誠にありがとうございます。

左右の反転というのはあまり聞かない現象ですが、トラックボールは正常でしょうか？
念のための確認ですが、右手側基板のLEFTジャンパーは半田付けなしということでOKでしょうか？

また原因を切り分けたいので、REMAPに登録しているファームウェアを【REMAP】から書き込んでみていただけますでしょうか。

お手数ですがご確認お願いいたします。

**Reply by monokaai** [reply]
> 左右の反転というのはあまり聞かない現象ですが、トラックボールは正常でしょうか？ 念のための確認ですが、右手側基板のLEFTジャンパーは半田付けなしということでOKでしょうか？
> また原因を切り分けたいので、REMAPに登録しているファームウェアを【REMAP】から書き込んでみていただけますでしょうか。

右手側ジャンパには半田付けしていませんでした。トラックボールですがREMAP上に
ご登録のファームウェアを書き込んだ場合は手前=下、奥=上、左=左...というマッピングですが
こちらもQMKを利用すると反転します。

まとめるとREMAPではキーマップ、トラックボールとも正常ですが
QMKだといずれも反転してしまいます。

**Reply by monokaai** [reply]
@Yowkees 
koron様の回答で無事解決に至りました。
この度はありがとうございました。今後ともよろしくお願いいたします。

**Comment by koron** [comment, marked as answer]
こちらに書いてある通りKeyballのファームウェアはQMK 0.22.14を使ってビルドすることを想定しています。
またビルド方法も同じドキュメントに書いてありますので参考にしてください。

https://github.com/Yowkees/keyball/blob/main/qmk_firmware/keyboards/keyball/readme.md#how-to-build

ローカルに環境を作らなくても、
レポジトリをフォーク・ブランチして書き換えcommit & pushすれば
GitHub Actionsに独自のファームをビルドさせることもできます。
このあたりはまだドキュメント化や使い勝手の面では不十分ですが、
よければご検討ください。

最新のQMK(0.24.x以降)と組み合わせる場合は1点不具合が報告されていますが、
こちらのPRにより対応可能らしいです。
https://github.com/Yowkees/keyball/pull/496

**Reply by monokaai** [reply]
QMKバージョンと不具合についてのコメントありがとうございます！
また[GitHub Actionsについても早々に実装＆ドキュメント化](https://github.com/Yowkees/keyball/discussions/548)して頂いたようで
ありがたい限りです。
こちらフォークしてのビルド＆書き込みもこれから試してみたいと思います。

**Reply by monokaai** [reply]
@koron 実装いただいたActionsでビルドされたファームウェアを
QMK Toolbox経由で書き込むことで正常な動作が確認できました！

QMKで変更したかったTAPPING_TERMも無事に変更でき、やりたかったことが達成できました。
この度は多大なご協力ありがとうございました。
